File: openweathermap-connector-scala/src/com/openweathermap/kafka/connect/WeatherAPIConnector.scala
package com.openweathermap.kafka.connect

import java.util.Arrays

import java.util.HashMap
import java.util.List
import java.util.Map
import java.util.stream.Collectors
import org.apache.kafka.common.config.ConfigDef
import org.apache.kafka.connect.connector.Task
import org.apache.kafka.connect.source.SourceConnector
import org.apache.kafka.connect.util.ConnectorUtils
import com.openweathermap.kafka.connect.WeatherAPIConfig.CITIES
import scala.collection.JavaConversions._

class WeatherAPIConnector extends SourceConnector {

  private var config: WeatherAPIConfig = _

  override def version(): String = "1.0"

  override def start(props: Map[String, String]): Unit = {
    config = new WeatherAPIConfig(props)
  }

  override def taskClass(): Class[_ <: Task] = classOf[WeatherAPITask]

  override def taskConfigs(maxTasks: Int): List[Map[String, String]] = {
    val cities: List[String] =
      Arrays.stream(config.getCities.split(",")).collect(Collectors.toList())
    val numGroups: Int = Math.min(cities.size, maxTasks)
    ConnectorUtils
      .groupPartitions(cities, numGroups)
      .stream()
      .map((taskCities) => {
        val taskProps: Map[String, String] =
          new HashMap[String, String](config.originalsStrings())
        taskProps.put(CITIES, String.join(",", taskCities))
        taskProps
      })
      .collect(Collectors.toList())
  }

  override def stop(): Unit = {}

  override def config(): ConfigDef = WeatherAPIConfig.config()

}

File: openweathermap-connector-java/src/com/openweathermap/kafka/connect/WeatherAPIConnector.java
package com.openweathermap.kafka.connect;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.connect.connector.Task;
import org.apache.kafka.connect.source.SourceConnector;
import org.apache.kafka.connect.util.ConnectorUtils;

import static com.openweathermap.kafka.connect.WeatherAPIConfig.CITIES;

public class WeatherAPIConnector extends SourceConnector {

	private WeatherAPIConfig config;

	@Override
	public String version() {
		return "1.0";
	}

	@Override
	public void start(Map<String, String> props) {
		config = new WeatherAPIConfig(props);
	}

	@Override
	public Class<? extends Task> taskClass() {
		return WeatherAPITask.class;
	}

	@Override
	public List<Map<String, String>> taskConfigs(int maxTasks) {
		List<String> cities = Arrays.stream(config.getCities().split(","))
				.collect(Collectors.toList());

		int numGroups = Math.min(cities.size(), maxTasks);

		return ConnectorUtils.groupPartitions(cities, numGroups)
				.stream()
				.map(taskCities -> {
					Map<String, String> taskProps = new HashMap<>(config.originalsStrings());
					taskProps.put(CITIES, String.join(",", taskCities));
					return taskProps;
				})
				.collect(Collectors.toList());
	}

	@Override
	public void stop() {
		// TODO Auto-generated method stub

	}

	@Override
	public ConfigDef config() {
		return WeatherAPIConfig.config();
	}

}